py/client/pydeephaven/session.py

```diff
@@ -327,49 +331,76 @@ def _fetch_fields(self):
             return resp.created if resp.created else []
 
     def _connect(self):
+        trace(f'_connect id={id(self)}')
         with self._r_lock:
+            if self.is_connected:
+                return
+            trace(f'_connect id={id(self)} connecting.')
             try:
                 scheme = "grpc+tls" if self._use_tls else "grpc"
                 self._flight_client = paflight.FlightClient(
                     location=f"{scheme}://{self.host}:{self.port}",
-                    middleware=[_DhClientAuthMiddlewareFactory(self)],
                     tls_root_certs=self._tls_root_certs,
                     cert_chain=self._client_cert_chain,
                     private_key=self._client_private_key,
                     generic_options=self._client_opts
                 )
-                self._auth_handler = _DhClientAuthHandler(self)
-                self._flight_client.authenticate(self._auth_handler)
             except Exception as e:
                 raise DHError("failed to connect to the server.") from e
 
             self.grpc_channel = self.session_service.connect()
 
+            # This RPC will get is the configuration and will also bootstrap
+            # our authentication to the server by virtue of sending the right
+            # header: "authorization" header key and our selected header value.
+            # The implementation will process the initial headers coming back
+            # from the server which will contain the bearer token we will
+            # use in subsequent RPCs; the token will be included in the updated
+            # value for self._auth_header_value that will happen through a call
+            # to update_metadata.
             config_dict = self.config_service.get_configuration_constants()
             session_duration = config_dict.get("http.session.durationMs")
             if not session_duration:
                 raise DHError("server configuration is missing http.session.durationMs")
 
             self.is_connected = True
 
-            self._timeout = int(session_duration.string_value)
+            self._timeout_seconds = int(session_duration.string_value)/1000.0
             if self._never_timeout:
                 self._keep_alive()
 
     def _keep_alive(self):
-        if self.is_connected:
-            if self._keep_alive_timer:
-                self._refresh_token()
-            self._keep_alive_timer = threading.Timer(self._timeout / 2 / 1000, self._keep_alive)
-            self._keep_alive_timer.daemon = True
-            self._keep_alive_timer.start()
-
-    def _refresh_token(self):
+        trace(f'_keep_alive')
+        if not self.is_connected:
+            return
+        factor = 1
```

Comment:
If factor is 1, won't timer_wakeup == self._timeout_seconds, thus we won't try a reconnect until the session has potentially already expired? I think under normal "keep alive" circumstances we want to refresh when the session around half finished, else it seems like we're trying to get lucky with at least one other gRPC call going off in the meantime?

If connection is lost, and factor is 1, and we just missed our chance to renew, what's the point in using factor = 0.5? If we had 5 mins to reconnect, and we failed, we should try a lot more aggressively to reconnect than 2.5 minutes.

I've probably misunderstood something here, because this doesn't make sense - I think the default factor should be something like .2 to .5, and some kind of exponential backoff (with noise) if we did fail to renew. That would also suggest a much larger _MAX_REFRESH_CONSECUTIVE_FAILURES than 3.

All that said, I think it might make sense to punt on this to a later PR, and just stick with _timeout_seconds/2 for now.

Reply:
Yeah I lost /2 everywhere there.